src/cts/src/HTreeBuilder.cpp

```diff
@@ -363,103 +346,463 @@ void adjustToplevelTopology(Point<double>& a,
+  candidates.emplace_back(newLoc);  // trial y#1
+
+  newY = py + targetDist - abs(px - newX);
+  newLoc.setY(newY);
+  candidates.emplace_back(newLoc);  // trial y#2
+
+  // move along x axis within blockage
+  newY = y;
+  newX = px - targetDist + abs(py - newY);
+  newLoc.setX(newX);
+  newLoc.setY(newY);
+  candidates.emplace_back(newLoc);  // trial x#1
+
+  newX = px + targetDist - abs(py - newY);
+  newLoc.setX(newX);
+  candidates.emplace_back(newLoc);  // trial x#2
+
+  for (const Point<double>& candidate : candidates) {
+    checkLegalityAndCostSpecial(currLoc,
+                                candidate,
+                                parentPoint,
+                                targetDist,
+                                sinks,
+                                scalingFactor,
+                                x1,
+                                y1,
+                                x2,
+                                y2,
+                                bestLoc,
+                                sinkDist,
+                                bestSinkDist);
+  }
+
+  if (bestLoc != noBest) {
+    return true;
+  }
   return false;
 }
 
-void findLegalPlacement(
-    const Point<double>& pt,
-    unsigned leng,
+void HTreeBuilder::checkLegalityAndCostSpecial(
+    Point<double> oldLoc,
+    Point<double> newLoc,
+    Point<double> parentPoint,
+    double targetDist,
+    const std::vector<Point<double>>& sinks,
+    int scalingFactor,
     double x1,
     double y1,
     double x2,
     double y2,
-    std::vector<Point<double>>& points  // candidate new locations
-)
+    Point<double>& bestLoc,
+    double& sinkDist,
+    double& bestSinkDist)
 {
-  double px = pt.getX();
-  double py = pt.getY();
-  std::vector<Point<double>> temp;
-  for (int i = 0; i < 2; ++i) {
-    double x = (i == 0) ? x1 : x2;
-    double y = (i == 0) ? y1 : y2;
-    double dx = leng - abs(px - x);
-    double dy = leng - abs(py - y);
-    if (x >= px - leng && x <= px + leng) {
-      temp.emplace_back(x, py + dx);
-      temp.emplace_back(x, py - dx);
-    }
-    if (y >= py - leng && y <= py + leng) {
-      temp.emplace_back(px + dy, y);
-      temp.emplace_back(px - dy, y);
+  if (fuzzyEqual(newLoc.computeDist(parentPoint), targetDist)
+      && checkLegalitySpecial(newLoc, x1, y1, x2, y2, scalingFactor)) {
+    sinkDist = weightedDistance(newLoc, oldLoc, sinks);
+    if (sinkDist < bestSinkDist) {
+      bestLoc = newLoc;
+      bestSinkDist = sinkDist;
     }
+    // clang-format off
+    debugPrint(logger_, CTS, "legalizer", 3, "adjustBestLegalLoc: branchPt move:{}=>{} is legal, "
+               "dist={:0.3f}, sinkDist={:0.3f}", oldLoc, newLoc, targetDist, sinkDist);
+    // clang-format on
+  }
+  // clang-format off
+  debugPrint(logger_, CTS, "legalizer", 3, "adjustBestLegalLoc: branchPt move:{}=>{} is illegal or dist "
+	     "{:0.3f} != {:0.3f}", oldLoc, newLoc, newLoc.computeDist(parentPoint), targetDist);
+  // clang-format off
```

Comment:
This messed up and left clang-format off not on.  Another reason to avoid such.

Reply:
Fixed.